presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java

```diff
@@ -145,10 +143,23 @@ public Result apply(TableScanNode scanNode, Captures captures, Context context)
                 // symbol should be mapped in redirected table scan
                 return symbol;
             }
-            // Column pruning after PPD into table scan can remove assignments for filter columns from the scan node
+
+            // validate that redirected types match source types
             Type domainType = requiredFilter.getDomains().get().get(destinationColumn).getType();
-            symbol = context.getSymbolAllocator().newSymbol(destinationColumn, domainType);
             ColumnHandle destinationColumnHandle = destinationColumnHandles.get(destinationColumn);
+            Type redirectedType = metadata.getColumnMetadata(context.getSession(), destinationTableHandle.get(), destinationColumnHandle).getType();
+            if (!domainType.equals(redirectedType)) {
```

Comment:
why validate again? is it guaranteed by `sourceType.equals(redirectedType))` above?

but this is good catch anyone, would be great to have not only a unit test, but also an integration test where redirect occurs after predicate pushdown, and query completes correctly. This would be very useful especially should we ever relax the `sourceType.equals(redirectedType))` condition above

Reply:
@findepi I think we're repeating the check here because in the earlier piece of code it is looking at table scan node assignments. Here we are looking at columns which are used in a filter but their assignment has been pruned from table can node.

Reply:
@findepi the check is needed. Consider scenario:
```
Output(a)
|
Filter(b>10)
|
TS(a,b)
```
then predicate is pushed into TS:
```
Output(a)
|
TS(a,b)
```
then column pruning happens:
```
Output(a)
|
TS(a)
```
then redirection happens:
```
Output(a)
|
Filter(redirected_b>10)
|
TS(a,redirected_b)
```
`redirected_b` is a new symbol therefore type compatibility was not checked in `sourceType.equals(redirectedType))`

 > but this is good catch anyone, would be great to have not only a unit test, but also an integration test where redirect occurs after predicate pushdown, and query completes correctly.

There is such integration test: `io.prestosql.tests.TestTpchTableScanRedirection#testTableScanRedirection`.